### ISSUE TITLE
the inverse of an isomorphism does not compute

### ISSUE THREAD

**Issue by benediktahrens** [UniMath/UniMath#279]
The following lemma is not proved by reflexivity. Should it (whatever that means)?

Require Import UniMath.CategoryTheory.precategories.

Lemma is_z_iso_compute {C: precategory}{a b : C}
 (f: a --> b) (g : b --> a)
 (H : is_inverse_in_precat f g)
 :
 inv_from_iso
      (isopair f (is_iso_from_is_z_iso _ 
                  (tpair (fun x => is_inverse_in_precat f x ) g H))) 
 = g.
Proof.
  cbn.
  apply id_right.
Qed.

**Comment by DanGrayson** [comment]
It seems the answer is "no".  Since this issue has been idle for 5 months and is referred to in #362, I'm closing it.  Feel free to reopen it.

**Comment by benediktahrens** [comment]
I would like to consider this issue again, and am thus reopening it.

**Comment by nmvdw** [comment]
I think it would be convenient if it could be proven by reflexivity, but that would mainly be beneficial if more complicated expressions are considered.

However, it might take some effort to make this change. The reason why it the inverses do not always compute nicely, is because of how `is_iso f` is defined. It says that precomposition with `f` is an equivalence, and then one can compute the inverse by taking the pre-image of the identity arrow. In the given example, you have arrows `f` and `g` such that `f · g = id` and `g · f = id`, and then the inverse of precomposition with `f` is precomposition with `g`. As a result, the inverse of `f` computes to `g · id`. I think the solution to this would be swapping the definitions of `is_iso` and `is_z_isomorphism`.

Note that in our formalization of bicategories, we chose a definition of invertible 2-cell/adjoint equivalence in which the inverse is given, and as far as I remember, we never had such issues.

**Comment by benediktahrens** [comment]
@nmvdw : i agree entirely with your analysis. However, there is a reason why `is_iso` was preferred over `is_z_iso`: the former is a proposition even when the hom-types are not sets. 

I am now not sure that the lack of computation is a price worth paying for the added generality. I am not aware of any interesting results about precategories.

@peterlefanulumsdaine : do you have an opinion?

**Comment by nmvdw** [comment]
What about using a version in which the inverse is given together with a coherency similar to the one in half-adjoint equivalences? If the hom-types are sets, then those coherencies would be trivial, so it would not be a burden in those cases.

**Comment by peterlefanulumsdaine** [comment]
@benediktahrens: I agree — I don’t think “precategories” have turned out to be a very important level of generality.  I’d support a change, so that the default definition used would be (what’s currently called) `is_z_iso`.

Here are several possible outlines of fixes.  In all of them, I assume that the current `is_iso` gets renamed to something more descriptive (e.g. `is_compweq_iso`).  Then:

1. `is_iso` is still defined in the generality of precategories, but its definition is changed to something else propositional, whose comparison with `is_z_iso` computes better (e.g. to “half-adjoint z-iso” as @nmvdw suggests, or else to “with a left and a right inverse”, assuming that those really are each propositional — I think they should be, but haven’t thought it through carefully)
2. `is_iso` is redefined as `is_z_iso`, but in the generality of precategories (where it’s not propositional)
3. `is_iso` is redefined as `is_z_iso`, and only defined in the generality of categories-with-homsets.  Then the comparisons with other definitions may be given in the library, but don’t need to be used when working in categories-with-homsets.

Both (1) and (3) seem quite reasonable to me.  (2) seems a bad idea (misleading terminology), but seems worth mentioning for completeness.  (1) sounds like it might be less work than (3), I guess?

**Comment by benediktahrens** [comment]
@peterlefanulumsdaine : I agree with your analysis, and option (1) sounds most appealing to me. Thanks to @nmvdw for this suggestion.

**Comment by rmatthes** [comment]
If I understand you well, you want to replay the definition [https://github.com/UniMath/UniMath/blob/master/UniMath/MoreFoundations/Equivalences.v#L17](https://github.com/UniMath/UniMath/blob/master/UniMath/MoreFoundations/Equivalences.v#L17) with morphisms in place of functions. By the way, is there a reason why the notion "half adjoint equivalence" is nowhere mentioned in the UniMath library (neither in source code nor documentation)?

**Comment by DanGrayson** [comment]
The file `MoreFoundations/Equivalences.v` presents them.  Here is the definition:
```coq
Definition isEquivalence {X Y:Type} (f:X->Y) :=
  ∑ (g:Y->X) (p:∏ y, f(g y) = y) (q:∏ x, g(f x) = x), ∏ x, maponpaths f (q x) = p(f x).
```

I called them "equivalences".

**Comment by rmatthes** [comment]
And would you think of a definition of isos in a category like this? [https://github.com/rmatthes/UniMath/blob/adjequivforisoincats/UniMath/CategoryTheory/Core/Isos.v#L377](https://github.com/rmatthes/UniMath/blob/adjequivforisoincats/UniMath/CategoryTheory/Core/Isos.v#L377):
`Definition is_hainverse_in_precat {C : precategory} {a b : C} (f : a --> b) (g : b --> a) :=
  ∑ (H1 : f · g = identity a) (H2 : g · f = identity b),
  pathscomp0 (maponpaths (postcomp_with f) H1) (id_left f) =
  pathscomp0 (assoc' f g f) (pathscomp0 (maponpaths (precomp_with f) H2) (id_right f)).`

**Comment by DanGrayson** [comment]
It looks interesting to me and seems worthwhile trying out.

**Comment by rmatthes** [comment]
I now think that my tentative definition is too bicategorical: when trying to find isomorphisms, one has to assume that the precategory forms a bicategory where the two-cells are given by paths and the usual operations there.

**Comment by peterlefanulumsdaine** [comment]
I guess if we’re aiming for route (1) — refactoring `is_iso` as some different coherent definition in precats that makes the computation better — then before any refactoring, we should prove the chosen definition really is an hprop.  It’s not clear to me whether “is half-adjoint equivalence” in an arbitrary precategory will be an hprop — I guess this is what @rmatthes had in mind as the definition being “too bicategorical”.

Another possibility is “bi-invertible maps” (which I like since it avoids higher paths).  But I guess this will be a proposition if and only if the half-adjoint ones are.

Has anyone else thought this through enough to be sure whether these are hprops in arbitrary precats, or not?

**Comment by benediktahrens** [comment]
@peterlefanulumsdaine : I agree with your plan. I have not checked if any alternative definition of `is_iso` is a proposition in precategories.

**Comment by benediktahrens** [comment]
This issue is resolved, just like the related #362
